R/return_linter.R

```diff
@@ -94,80 +86,58 @@ return_linter <- function(
-    #   two top level branches have at least two return()s
-    # because of special 'in' syntax for 'for' loops, the condition is
-    #   tagged differently than for 'if'/'while' conditions (simple PAREN)
-    xpath <- glue("
+    body_xpath <- glue("
     (//FUNCTION | //OP-LAMBDA)[parent::expr[not(
       preceding-sibling::expr[SYMBOL[{ xp_text_in_table(except) }]]
     )]]
       /following-sibling::expr[OP-LEFT-BRACE and expr[last()]/@line1 != @line1]
       /expr[last()]
-      /*[
-        (
-          position() = 1
-          and (
-            (
-              { xp_or(paste0('self::', setdiff(control_calls, 'IF'))) }
-            ) or (
-              not({ xp_or(paste0('self::', control_calls)) })
-              and not(
-                following-sibling::PIPE
-                or following-sibling::SPECIAL[text() = '%>%']
-              )
-              and not(self::expr/SYMBOL_FUNCTION_CALL[
-                { xp_text_in_table(return_functions) }
-              ])
-            )
-          )
-        ) or (
-          preceding-sibling::IF
-          and self::expr
-          and position() > 4
-          and not(.//SYMBOL_FUNCTION_CALL[{ xp_text_in_table(return_functions) }])
-        )
-      ]
     ")
+    lint_xpath <- glue("self::*[not(
+      (self::expr | following-sibling::SPECIAL[text() = '%>%']/following-sibling::expr/expr[1])
+        /SYMBOL_FUNCTION_CALL[{ xp_text_in_table(return_functions) }]
+    )]")
     msg <- "All functions must have an explicit return()."
   }
 
   Linter(linter_level = "expression", function(source_expression) {
     xml <- source_expression$xml_parsed_content
     if (is.null(xml)) return(list())
 
-    xml_nodes <- xml_find_all(xml, xpath)
+    body_expr <- xml_find_all(xml, body_xpath)
+
+    # nested_return_lints not "vectorized" due to xml_children()
+    lapply(body_expr, nested_return_lints, lint_xpath, source_expression, msg)
+  })
+}
+
+nested_return_lints <- function(expr, xpath, source_expression, lint_message) {
+  child_expr <- xml_children(expr)
+  if (length(child_expr) == 0L) {
+    return(list())
+  }
+  child_node <- xml_name(child_expr)
 
-    xml_nodes_to_lints(
-      xml_nodes,
+  if (child_node[1L] == "OP-LEFT-BRACE") {
+    expr_idx <- which(child_node %in% c("expr", "equal_assign", "expr_or_assign_or_help"))
+    if (length(expr_idx) == 0L) { # empty brace expression {}
+      return(list())
+    }
+    nested_return_lints(child_expr[[tail(expr_idx, 1L)]], xpath, source_expression, lint_message)
+  } else if (child_node[1L] == "IF") {
+    expr_idx <- which(child_node %in% c("expr", "equal_assign", "expr_or_assign_or_help"))
+    c(
+      # TRUE condition
+      nested_return_lints(child_expr[[expr_idx[2L]]], xpath, source_expression, lint_message),
+      # FALSE condition, if present
+      if (length(expr_idx) > 2L) nested_return_lints(child_expr[[expr_idx[3L]]], xpath, source_expression, lint_message)
+    )
+  } else {
+    list(xml_nodes_to_lints(
```

Comment:
Why is `list()` needed?

Reply:
otherwise `c()` is used on the lints and they break, i.e. `c(xml_nodes_to_lints(...), xml_nodes_to_lints(...))` doesn't work. Not sure if it should but this is simple enough.

Reply:
(I guess that's not exactly true, we have plenty of examples of building lints like `c(lintsA, lintsB, lintsC)`. I guess it's the recursion that's breaking things in this case? What's certain is I originally tried without `list()` and it didn't work (I got back a list looking like `c(as.list(lintsA), as.list(lintsB))` with duplicate names and the wrong class), but does work with `list()`, so I'm content :)

Reply:
OK it bugged me. The issue comes down to the `c(...)` approach, which breaks when `else` is missing, e.g. `c(xml_nodes_to_lints(...), NULL)` strips the `lint` attribute, breaking things.

A solution is to take your advice from the [other comment](https://github.com/r-lib/lintr/pull/2362#discussion_r1408185541) and use `lapply()` instead of `c()`. It's actually much cleaner now that I've condensed the recurser signature, too. `unlist()` is not needed -- `flatten_lints()` handles that later on.